model/sharing/sharing.go

```diff
@@ -1056,206 +1056,315 @@ func CheckSharings(inst *instance.Instance) ([]map[string]interface{}, error) {
-						"id":     s.SID,
-						"type":   "missing_oauth_client",
-						"member": i,
-						"owner":  true,
-					})
-				}
-				if s.Credentials[i-1].AccessToken == nil {
-					checks = append(checks, map[string]interface{}{
-						"id":     s.SID,
-						"type":   "missing_access_token",
-						"member": i,
-						"owner":  true,
-					})
-				}
-				if m.Instance == "" {
-					checks = append(checks, map[string]interface{}{
-						"id":     s.SID,
-						"type":   "missing_instance_for_member",
-						"member": i,
-					})
-				}
-			}
+	return checks
+}
 
-			if len(s.Credentials)+1 != len(s.Members) {
-				checks = append(checks, map[string]interface{}{
-					"id":         s.SID,
-					"type":       "invalid_number_of_credentials",
-					"owner":      true,
-					"nb_members": len(s.Credentials),
-				})
-				return nil
-			}
-		} else {
-			if len(s.Credentials) != 1 {
-				checks = append(checks, map[string]interface{}{
-					"id":         s.SID,
-					"type":       "invalid_number_of_credentials",
-					"owner":      false,
-					"nb_members": len(s.Credentials),
-				})
-				return nil
-			}
+func (s *Sharing) checkSharingTreesConsistency(inst *instance.Instance) (checks []map[string]interface{}, err error) {
+	if !s.Owner || !s.Active || s.Initial {
+		return nil, nil
+	}
 
-			if s.Credentials[0].InboundClientID == "" {
-				checks = append(checks, map[string]interface{}{
-					"id":    s.SID,
-					"type":  "missing_inbound_client_id",
-					"owner": false,
-				})
-			}
+	rule := s.FirstFilesRule()
+	if rule == nil || !rule.HasSync() || !rule.FilesByID() {
+		// TODO: handle referenced_by type sharings?
+		return nil, nil
+	}
+	//sharedRoot := rule.Values[0] // This is the shared io.cozy.files id
 
-			if !s.ReadOnly() && s.Members[0].Instance == "" {
-				checks = append(checks, map[string]interface{}{
-					"id":     s.SID,
-					"type":   "missing_instance_for_member",
-					"member": 0,
-				})
-			}
+	var domains []string
+	for _, m := range s.Members {
+		if m.Status != MemberStatusReady {
+			continue
 		}
 
-		return nil
-	})
-	return checks, err
+		domains = append(domains, m.Instance)
+	}
+
+	if len(domains) == 0 {
+		return nil, nil
+	}
+
+	//var tree map[string]interface{}
+	//fn := func(name string, dir *vfs.DirDoc, file *vfs.FileDoc, err error) error {
+	//	tree[name] = map
+	//	return nil
+	//}
+	//fs := inst.VFS()
+	//if err := vfs.WalkByID(fs, sharedRoot, fn); err != nil {
+	//	return err
+	//}
+	ownersDocs, err := FindMatchingDocs(inst, *rule)
+	if err != nil {
+		checks = append(checks, map[string]interface{}{
+			"id":   s.SID,
+			"type": "missing_matching_docs_for_owner",
+			"err":  err,
+		})
+		return checks, nil
+	}
+
+	for _, d := range domains {
+		inst, err := instance.GetFromCouch(d)
+		if err != nil {
+			checks = append(checks, map[string]interface{}{
+				"id":     s.SID,
+				"type":   "missing_instance_for_member",
+				"member": d,
+			})
+		}
+
+		instDocs, err := FindMatchingDocs(inst, *rule)
```

Comment:
Does it work? I don't the transformation (XOR) on the id

Reply:
I missed that part. It should be good now.